R/Task.R

```diff
@@ -1056,23 +1071,69 @@ Task = R6Class("Task",
     },
 
     #' @field weights ([data.table::data.table()])\cr
-    #' If the task has a column with designated role `"weight"`, a table with two columns:
+    #' Deprecated, use `$weights_learner` instead.
+    weights = function(rhs) {
+      assert_ro_binding(rhs)
+      .Deprecated("Task$weights_learner", old = "Task$weights")
+      self$weights_learner
+    },
+
+    #' @field weights_learner ([data.table::data.table()])\cr
+    #' Returns the observation weights used for training a [Learner] (column role `weights_learner`)
+    #' as a `data.table` with the following columns:
     #'
     #' * `row_id` (`integer()`), and
-    #' * observation weights `weight` (`numeric()`).
+    #' * `weight` (`numeric()`).
     #'
-    #' Returns `NULL` if there are is no weight column.
-    weights = function(rhs) {
+    #' Returns `NULL` if there are is no column with the designated role.
+    weights_learner = function(rhs) {
+      assert_has_backend(self)
+      assert_ro_binding(rhs)
+      weight_cols = private$.col_roles[["weights_learner"]]
+      if (length(weight_cols) == 0L) {
+        return(NULL)
+      }
+      data = self$backend$data(private$.row_roles$use, c(self$backend$primary_key, weight_cols))
```

Comment:
This might fail because columns can be returned in arbitrary order: https://github.com/mlr-org/mlr3/blob/57b6109355ae5219d735a89c188d1b987ae2ff41/R/DataBackendDataTable.R#L61

Instead you should use the `$data()` method of the task: https://github.com/mlr-org/mlr3/blob/57b6109355ae5219d735a89c188d1b987ae2ff41/R/Task.R#L265